Orange/statistics/util.py

```diff
@@ -10,80 +10,181 @@
 import bottleneck as bn
 
 
-def _count_nans_per_row_sparse(X, weights):
+def _count_nans_per_row_sparse(X, weights, dtype=None):
     """ Count the number of nans (undefined) values per row. """
-    items_per_row = 1 if X.ndim == 1 else X.shape[1]
-    counts = np.ones(X.shape[0]) * items_per_row
-    nnz_per_row = np.bincount(X.indices, minlength=len(counts))
-    counts -= nnz_per_row
     if weights is not None:
-        counts *= weights
-    return np.sum(counts)
+        X = X.tocoo(copy=False)
+        nonzero_mask = np.isnan(X.data)
+        nan_rows, nan_cols = X.row[nonzero_mask], X.col[nonzero_mask]
+
+        if weights.ndim == 1:
+            data_weights = weights[nan_rows]
+        else:
+            data_weights = weights[nan_rows, nan_cols]
+
+        w = sp.coo_matrix((data_weights, (nan_rows, nan_cols)), shape=X.shape)
+        w = w.tocsr()
+
+        return np.fromiter((np.sum(row.data) for row in w), dtype=dtype)
+
+    return np.fromiter((np.isnan(row.data).sum() for row in X), dtype=dtype)
+
+
+def sparse_count_zeros(x):
+    """ Count the number of implicit zeros in a sparse matrix. """
+    if not sp.issparse(x):
+        raise TypeError('The matrix provided was not sparse.')
+    return np.prod(x.shape) - x.nnz
+
+
+def sparse_has_zeros(x):
+    """ Check if sparse matrix contains any implicit zeros. """
+    if not sp.issparse(x):
+        raise TypeError('The matrix provided was not sparse.')
+    return np.prod(x.shape) != x.nnz
 
 
-def bincount(X, max_val=None, weights=None, minlength=None):
+def sparse_zero_weights(x, weights):
```

Comment:
Is this only meant to be used when both `x` and `weights` are of the same shape, i.e. one dimensional?

Reply:
Adding a check here causes more problems than it's worth, because this can be called in two ways that are valid:
1. `x` has shape `(1, n)`
2. `x` has shape `(n, 1)`
These are basically equivalent since they are both "1d", but this would make the check more complicated.